File: python/fathomdb/_feedback.py
from __future__ import annotations

import threading
import time
import uuid
from collections.abc import Callable, Mapping
from typing import TypeVar

from ._types import FeedbackConfig, ResponseCycleEvent, ResponseCyclePhase

T = TypeVar("T")


def run_with_feedback(
    *,
    surface: str,
    operation_kind: str,
    metadata: Mapping[str, str] | None,
    progress_callback: Callable[[ResponseCycleEvent], None] | None,
    feedback_config: FeedbackConfig | None,
    operation: Callable[[], T],
) -> T:
    if progress_callback is None:
        return operation()

    config = feedback_config or FeedbackConfig()
    metadata_dict = dict(metadata or {})
    operation_id = uuid.uuid4().hex
    started_at = time.monotonic()
    stop_event = threading.Event()
    callback_disabled = False
    callback_lock = threading.Lock()

    def emit(
        phase: ResponseCyclePhase,
        *,
        error_code: str | None = None,
        error_message: str | None = None,
    ) -> None:
        nonlocal callback_disabled
        with callback_lock:
            if callback_disabled:
                return
            event = ResponseCycleEvent(
                operation_id=operation_id,
                operation_kind=operation_kind,
                surface=surface,
                phase=phase,
                elapsed_ms=int((time.monotonic() - started_at) * 1000),
                slow_threshold_ms=config.slow_threshold_ms,
                metadata=dict(metadata_dict),
                error_code=error_code,
                error_message=error_message,
            )
            try:
                progress_callback(event)
            except Exception:
                callback_disabled = True

    def heartbeat_loop() -> None:
        if stop_event.wait(config.slow_threshold_ms / 1000):
            return
        emit(ResponseCyclePhase.SLOW)
        while not stop_event.wait(config.heartbeat_interval_ms / 1000):
            emit(ResponseCyclePhase.HEARTBEAT)

    emit(ResponseCyclePhase.STARTED)
    timer_thread = threading.Thread(target=heartbeat_loop, daemon=True)
    timer_thread.start()

    try:
        result = operation()
    except Exception as error:
        stop_event.set()
        timer_thread.join()
        emit(
            ResponseCyclePhase.FAILED,
            error_code=type(error).__name__,
            error_message=str(error),
        )
        raise

    stop_event.set()
    timer_thread.join()
    emit(ResponseCyclePhase.FINISHED)
    return result

File: archive/pre-0.6.0/typescript/packages/fathomdb/src/feedback.ts
import { randomUUID } from "node:crypto";
import { performance } from "node:perf_hooks";
import type { FeedbackConfig, ProgressCallback, ResponseCycleEvent, ResponseCyclePhase } from "./types.js";

/** Surface identifier included in feedback events to distinguish the TypeScript SDK. */
export const SURFACE = "typescript";

const DEFAULT_SLOW_THRESHOLD_MS = 500;

/**
 * Wrap a synchronous operation with response-cycle feedback events.
 *
 * If `progressCallback` is undefined, the operation is called directly.
 * Otherwise emits STARTED before the call, and FINISHED or FAILED after.
 *
 * Note: Since native engine calls are synchronous and block the Node.js
 * event loop, SLOW/HEARTBEAT events cannot fire during the operation.
 * Only STARTED and FINISHED/FAILED are emitted in the current synchronous
 * binding. The timer infrastructure will activate once async bindings
 * are available.
 *
 * @param opts - Feedback wrapper options.
 * @param opts.operationKind - Identifies the operation (e.g. `"engine.open"`, `"query.execute"`).
 * @param opts.metadata - Arbitrary key-value metadata included in every emitted event.
 * @param opts.progressCallback - Optional callback invoked with {@link ResponseCycleEvent} instances.
 * @param opts.feedbackConfig - Timing thresholds controlling slow/heartbeat emission.
 * @param opts.operation - The synchronous operation to execute.
 * @returns The return value of `opts.operation`.
 */
export function runWithFeedback<T>(opts: {
  operationKind: string;
  metadata: Record<string, string>;
  progressCallback: ProgressCallback | undefined;
  feedbackConfig: FeedbackConfig | undefined;
  operation: () => T;
}): T {
  if (!opts.progressCallback) {
    return opts.operation();
  }

  const callback = opts.progressCallback;
  const slowThresholdMs = opts.feedbackConfig?.slowThresholdMs ?? DEFAULT_SLOW_THRESHOLD_MS;
  const operationId = randomUUID().replace(/-/g, "");
  const startedAt = performance.now();
  let callbackDisabled = false;

  function emit(phase: ResponseCyclePhase, errorCode?: string, errorMessage?: string): void {
    if (callbackDisabled) return;
    const event: ResponseCycleEvent = {
      operationId,
      operationKind: opts.operationKind,
      surface: SURFACE,
      phase,
      elapsedMs: Math.round(performance.now() - startedAt),
      slowThresholdMs,
      metadata: opts.metadata,
      errorCode,
      errorMessage,
    };
    try {
      callback(event);
    } catch {
      callbackDisabled = true;
    }
  }

  emit("started");
  try {
    const result = opts.operation();
    emit("finished");
    return result;
  } catch (error) {
    const name = error instanceof Error ? error.constructor.name : "Error";
    const message = error instanceof Error ? error.message : String(error);
    emit("failed", name, message);
    throw error;
  }
}
